### PR TITLE
Finish the book challenge

### PR DESCRIPTION
**Pull request by AmberWilkie**

### CODE REVIEW

File: challenge_5/book.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ def initialize(attrs)
+   self.author = attrs[:author]
+   self.title = attrs[:title]
+   self.publisher = attrs[:publisher]
+   self.year = attrs[:year]
+ end
+
+ def citate(page)
+   "#{self.author}, #{self.title}, #{self.publisher}, #{self.year}, #{page}"
+ end
```
Comment (tochman) [inline challenge_5/book.rb]: So what would happen if I try to cite page 200 in a 150 pages book?

**Comment by AmberWilkie** [comment]
I still think the instructions are unclear. Here's how I would write this:

Using Test Driven Development, construct a ruby Class named “Book”
- add the following attributes: “title”, “author”, “publishing_year”, “publisher”
- add an instance method that returns a source citation of the book, noting which page we are referencing.
  i.e. “Thomas Ochman, Coding As A Craft, Pragmatic Sweden, 2016, 50”
- your book should have 200 total pages

Create a folder structure of your choosing in this folder and save your work.

**Comment by AmberWilkie** [comment]
Also, yes, it's cite and not citate. I no speak well English.

File: challenge_5/book.rb [hunk 2]
```diff
@@ -10,7 +10,11 @@ def initialize(attrs)
  end
 
  def citate(page)
-   "#{self.author}, #{self.title}, #{self.publisher}, #{self.year}, #{page}"
+   if page < 999
```
Comment (tochman) [inline challenge_5/book.rb]: Where does 999 come from? Why not use `self.pages` / `@pages`?

**Comment by AmberWilkie** [comment]
Yeah I still didn't get it. And then I error-handled everything I could think of. Learned about guard clauses, so that was fun.

File: challenge_5/book.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
+class Book
+  attr_accessor :author, :title, :publisher, :year, :total_pages, :citation
+
+  def initialize(attrs)
+    self.author = attrs[:author]
```
Comment (diraulo) [inline challenge_5/book.rb:5]: any particular reason why you prefer `self.` over `@` here?

File: challenge_5/book.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    self.title = attrs[:title]
+    self.publisher = attrs[:publisher]
+    self.year = attrs[:year]
+    self.total_pages = attrs[:total_pages]
+  end
+
+  def citate(attrs)
```
Comment (diraulo) [inline challenge_5/book.rb:12]: will this method receive any other attribute than the page number?

File: challenge_5/book.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    @page = attrs[:page]
+    handle_no_page_number
+    handle_whacky_page_input
+    raise 'no author found' if author.nil?
+    raise 'no title found' if title.nil?
+    if page_checks_out?
```
Comment (diraulo) [inline challenge_5/book.rb:18]: you could make use of a guard clause here too.

``` ruby
raise 'Get Bent' unless page_checks_out?
create_citation
```

This way if no exception is raised we know all is good and we can create the citation.

File: challenge_5/book.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      raise 'please enter a page number'
+    end
+  end
+
+  def page_checks_out?
+    (
+    (@page == 'no page given')) ||
+      (
+      (@page.is_a? Integer) &&
+      (@page <= total_pages)
```
Comment (diraulo) [inline challenge_5/book.rb:48]: what if `@page` is a negative number?

File: challenge_5/book.rb (hunk 3 above)
Comment (AmberWilkie) [reply challenge_5/book.rb:5]: I honestly don't have a good handle on self.variable vs. @variable.

File: challenge_5/book.rb (hunk 4 above)
Comment (AmberWilkie) [reply challenge_5/book.rb:12]: I had to set it up like that if I wanted to deal with someone entering no page number. From what I remember of college, you would sometimes cite an entire book or sometimes you are citing news articles that maybe don't have pages, so you'd have nothing to put in the "page" area. I don't have any _use case_ for this program, so I'm just doing whatever, basically.

File: challenge_5/book.rb (hunk 5 above)
Comment (AmberWilkie) [reply challenge_5/book.rb:18]: Cool. Makes sense.

File: challenge_5/spec/book_spec.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
+require './book'
+
+describe Book do
+
+  before do
```
Comment (diraulo) [inline challenge_5/spec/book_spec.rb:5]: consider using `let` over `before` blocks 👉 http://betterspecs.org/#let

File: challenge_5/book.rb (hunk 4 above)
Comment (diraulo) [reply challenge_5/book.rb:12]: I hear you but the **requirements** in this context are about citing a given page.

But let's say we consider a scenario where `page` is not provided. You could set an attribute with default value of maybe `nil` and handle that case in the method

Comment (AmberWilkie) [inline challenge_5/book.rb:12]: I would argue the requirements of this challenge allow for consideration of
no input. I don't know what form the "no input" would take, so I tried to
handle that. But thanks :)

_Amber Wilkie_
www.amberwilkie.com
0720.08.27.07 | Instagram http://www.instagram.com/amberwphoto

**Pull request closed without merging**